<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: 24.06.2026

----



##**Бизнес-контекст:**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных:**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Шаг 1. Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (Date в datetime, CustomerNo в int).
- Проверка и обработка пропусков (CustomerNo).
- Проверка дубликатов и выбросов (Price, Quantity).


---
## **Библиотеки**




In [2]:

# Необходимые библиотеки

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    from scipy.stats import ttest_ind
    from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn scipy matplotlib-venn kagglehub -q
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    from scipy.stats import ttest_ind
    from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


# // Оптимизировать загрузку библиотек ^^^^^

## **Загрузка датасета:**

In [3]:
# ============================================
# ШАГ 2: Скачиваем датасет через kagglehub
# ============================================
print("⏳ Скачивание датасета...")
path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
print(f"✅ Путь к данным: {path}")

⏳ Скачивание датасета...
Using Colab cache for faster access to the 'an-online-shop-business' dataset.
✅ Путь к данным: /kaggle/input/an-online-shop-business


In [4]:

# ============================================
# ШАГ 3: Смотрим все файлы в папке
# ============================================
print("\n📁 Найденные файлы:")
all_files = os.listdir(path)
for file in all_files:
    print(f"  - {file}")


📁 Найденные файлы:
  - Sales Transaction v.4a.csv


In [5]:
# ============================================
# ШАГ 4: Загружаем CSV файл
# ============================================
# Ищем все CSV файлы
csv_files = [f for f in all_files if f.endswith('.csv')]

if csv_files:
    # Берем первый CSV файл (это и есть ваш Sales Transaction v.4a.csv)
    file_name = csv_files[0]
    file_path = os.path.join(path, file_name)

    print(f"\n📂 Загружаем файл: {file_name}")
    df = pd.read_csv(file_path)


📂 Загружаем файл: Sales Transaction v.4a.csv


#// упростить загрузку ДАТАСЕТА ^^^^^

## **Осмотр данных**

In [6]:
df.columns

Index(['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Price',
       'Quantity', 'CustomerNo', 'Country'],
      dtype='object')

In [34]:
df.shape

(536350, 8)

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [30]:
df.head(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,2019-12-09,22485,Set Of 2 Wooden Market Crates,21.47,12,17490,United Kingdom
1,581475,2019-12-09,22596,Christmas Star Wish List Chalkboard,10.65,36,13069,United Kingdom
2,581475,2019-12-09,23235,Storage Tin Vintage Leaf,11.53,12,13069,United Kingdom
3,581475,2019-12-09,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069,United Kingdom
4,581475,2019-12-09,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069,United Kingdom
5,581475,2019-12-09,21705,Bag 500g Swirly Marbles,10.65,24,13069,United Kingdom
6,581475,2019-12-09,22118,Joy Wooden Block Letters,11.53,18,13069,United Kingdom
7,581475,2019-12-09,22119,Peace Wooden Block Letters,12.25,12,13069,United Kingdom
8,581475,2019-12-09,22217,T-Light Holder Hanging Lace,10.65,12,13069,United Kingdom
9,581475,2019-12-09,22216,T-Light Holder White Lace,10.55,24,13069,United Kingdom


In [35]:
df.sample(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
450455,543631,2019-02-10,21485,Retrospot Heart Hot Water Bottle,15.32,2,14081,United Kingdom
488760,540468,2019-01-07,21383,Pack Of 12 Sticky Bunnies,11.95,1,12468,Germany
317747,556202,2019-06-09,90214K,"Letter ""K"" Bling Key Ring",11.53,3,14606,United Kingdom
478971,541202,2019-01-14,21677,Hearts Stickers,11.12,12,17581,United Kingdom
262443,561218,2019-07-25,22615,Pack Of 12 Circus Parade Tissues,11.10,3,15218,United Kingdom
63915,577039,2019-11-17,23581,Jumbo Bag Paisley Park,6.19,10,12539,Spain
13561,580632,2019-12-05,85178,Victorian Sewing Kit,6.04,3,16360,United Kingdom
25388,579524,2019-11-30,23300,Gardeners Kneeling Pad Cup Of Tea,6.39,12,16295,United Kingdom
366802,551412,2019-04-28,21980,Pack Of 12 Red Retrospot Tissues,10.55,12,15508,United Kingdom
311693,556932,2019-06-15,20846,Zinc Heart Lattice T-Light Holder,12.77,1,14932,United Kingdom


In [46]:
df.describe()

,Date,Price,Quantity,CustomerNo
count,536350,536350.000000,536350.000000,536350.0
mean,2019-07-04 02:52:31.891116032,12.662182,9.919347,15226.331532
min,2018-12-01 00:00:00,5.130000,-80995.000000,-1.0
25%,2019-03-28 00:00:00,10.990000,1.000000,13804.0
50%,2019-07-20 00:00:00,11.940000,3.000000,15152.0
75%,2019-10-19 00:00:00,14.090000,10.000000,16729.0
max,2019-12-09 00:00:00,660.620000,80995.000000,18287.0
std,NaN,8.490450,216.662300,1723.407831


### **Анализ уникальных позиций**

In [10]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,23204
CustomerNo,4738
ProductName,3768
Country,38


### **Анализ хвоста**

In [11]:
df['Price'].quantile([0.25, 0.5, 0.75, 0.9, 0.95, 0.99])

,Price
0.25,10.99
0.50,11.94
0.75,14.09
0.90,17.37
0.95,19.42
0.99,27.30


### **В результате предварительно осмотра установлено:**

- Объём выборки: 536 350 записей, 8 признаков.
- Пропуски: 0.01%  `CustomerNo`
- Размерность данных: 23 204 транзакции, 4 738 клиентов, 3 768 товаров, 38 стран.
- Распределение цен: медиана — 11.94 руб., 90-й перцентиль — 17.37 руб., 99-й — 27.30 руб.
- Выбросы по цене (> 27.30 руб.) составляют 1% выборки.

 Данные пригодны для дальнейшего исследовательского анализа

## **Обработка  типов данных**

In [12]:
#осмотр данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [13]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


# // добавить коммент

## **Поиск и обработка пропусков**

In [14]:
# Пропуски: количество и доля
df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [39]:
df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


In [40]:
# Заменить NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

In [43]:
# ============================================
# ПРОВЕРКА ЗАМЕНЫ ПРОПУСКОВ
# ============================================

# 1. Найти строки, где были пропуски (до замены)
print("="*50)
print("СТРОКИ, ГДЕ БЫЛИ ПРОПУСКИ:")
print("="*50)
display(df[df['CustomerNo'] == -1].head(10))

# 2. Проверить, сколько стало -1
print(f"\nВсего замен: {(df['CustomerNo'] == -1).sum()}")

СТРОКИ, ГДЕ БЫЛИ ПРОПУСКИ:


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom



Всего замен: 55


# // краткий коммент

# **// Понять природу отрицательного числа**

## **Поиск и дубликатов/неявных дубликатов**

In [15]:
df.duplicated().sum()

np.int64(5200)

# // краткий коммент

#**Шаг 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

# **Шаг 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

In [15]:
#топ-10 дорогих товаров

top_price = df.groupby('ProductName')['Price'].max().sort_values(ascending=False).head(10)
print(top_price)

# **Шаг 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.